spacy/util.py

```diff
@@ -256,20 +258,85 @@ def lang_class_is_loaded(lang: str) -> bool:
     return lang in registry.languages
 
 
+def find_matching_language(lang: str) -> Optional[str]:
+    """
+    Given an IETF language code, find a supported spaCy language that is a
+    close match for it (according to Unicode CLDR language-matching rules).
+    This allows for language aliases, ISO 639-2 codes, more detailed language
+    tags, and close matches.
+
+    Returns the language code if a matching language is available, or None
+    if there is no matching language.
+
+    >>> find_matching_language('en')
+    'en'
+    >>> find_matching_language('pt-BR')  # Brazilian Portuguese
+    'pt'
+    >>> find_matching_language('fra')  # an ISO 639-2 code for French
+    'fr'
+    >>> find_matching_language('iw')  # obsolete alias for Hebrew
+    'he'
+    >>> find_matching_language('no')  # Norwegian
+    'nb'
+    >>> find_matching_language('mo')  # old code for ro-MD
+    'ro'
+    >>> find_matching_language('zh-Hans')  # Simplified Chinese
+    'zh'
+    >>> find_matching_language('zxx')
+    None
+    """
+    import spacy.lang  # noqa: F401
+    if lang == 'xx':
+        return 'xx'
+
+    # Find out which language modules we have
+    possible_languages = []
+    for modinfo in pkgutil.iter_modules(spacy.lang.__path__):
+        code = modinfo.name
+        if code == 'xx':
+            # Temporarily make 'xx' into a valid language code
+            code = 'mul'
+        try:
+            language = langcodes.Language.get(code)
+            if language.is_valid():
+                possible_languages.append(code)
+        except langcodes.tag_parser.LanguageTagError:
+            pass
+
+    match, _dist = langcodes.closest_match(
+        lang, possible_languages, max_distance=10
+    )
+    if match == 'mul':
+        # Convert 'mul' back to spaCy's 'xx'
+        return 'xx'
+    elif match != 'und':
+        return match
```

Comment:
It's a little unfortunate that we need to check for a hardcoded string value of another library. I don't suppose `"und"` is defined in a variable somewhere? Why doesn't `closest_match` use `None` in these cases?

Reply:
In langcodes I wanted to return something that can be parsed as a language code instead of adding an optional type, and 'und' is the BCP 47 code for a language code that provides no information.

A different way to get this string value from langcodes would be `str(langcodes.Language(language=None))`. I could also add that as a constant in the next version (it's due for one when CLDR v40 comes out in early October).

Reply:
...in fact, looking back, there was a standard that said I should return 'und' in that case.

But I understand why a function with a different signature would help. Maybe I should add some functions to a version of langcodes now, without waiting for the data to change.